common/replay/replay_guardian_impl.go

```diff
@@ -0,0 +1,166 @@
+
+	// A queue of observed hashes, ordered by expiration time. Used to prune old hashes.
+	expirationQueue *priorityqueue.Queue
+
+	// A mutex to protect the observedHashes and expirationQueue.
+	lock sync.Mutex
+}
+
+// hashWithTimestamp is a request hash with self-reported timestamp associated with that request.
+type hashWithTimestamp struct {
+	hash      string
+	timestamp time.Time
+}
+
+// NewReplayGuardian creates a new ReplayGuardian.
+//
+// maxTimeInFuture is the maximum amount of time that a request's timestamp can be ahead of the local wall clock time.
+// Increasing this value permits more leniency in the timestamp of incoming requests, at the potential cost of a higher
+// memory overhead. In theory, if requests are sent with a timestamp exactly at the maximum time in the future,
+// this utility will remember them for a total of (maxTimeInFuture + maxTimeInPast), since that is the amount of time
+// that will need to elapse locally before the request exceeds the maximum age. If maxTimeInFuture is extremely large,
+// then an attacker may be able to cause this utility to be forced to remember a very large amount of data.
+//
+// maxTimeInPast is the maximum amount of time that a request's timestamp can be behind the local wall clock time.
+// Increasing this value permits more leniency in the timestamp of incoming requests, at the potential cost of a higher
+// memory overhead.
+func NewReplayGuardian(
+	timeSource func() time.Time,
+	maxTimeInPast time.Duration,
+	maxTimeInFuture time.Duration) ReplayGuardian {
+
+	return &replayGuardian{
+		timeSource:      timeSource,
+		maxTimeInFuture: maxTimeInFuture,
+		maxTimeInPast:   maxTimeInPast,
+		observedHashes:  make(map[string]struct{}),
+		expirationQueue: priorityqueue.NewWith(compareHashWithTimestamp),
+	}
+}
+
+// compareKeyWithExpiration compares two hashWithTimestamp objects by their expiration time. Used to create
+// a priority queue that orders the requests in chronological order (i.e. the order in which they will expire).
+func compareHashWithTimestamp(a interface{}, b interface{}) int {
+
+	keyA := a.(*hashWithTimestamp)
+	keyB := b.(*hashWithTimestamp)
+
+	if keyA.timestamp.Before(keyB.timestamp) {
+		return -1
+	} else if keyA.timestamp.After(keyB.timestamp) {
+		return 1
+	}
+	return 0
+}
+
+// VerifyRequest verifies that a request with the given hash and timestamp is not a replay
+// of a previous request. If it cannot be determined if a request is a replay or not,
+// then the request is rejected. Only if it can be guaranteed that the request is not a replay
+// will this method return nil.
+//
+// In order to be a verified unique request, the following conditions must be met:
+// - the request's timestamp must be no more than X minutes ahead of the local wall clock time
+// - the request's timestamp must be no more than Y minutes behind the local wall clock time
+// - the request's hash must not have been previously observed (hashes are remembered for Y minutes)
```

Comment:
> (hashes are remembered for Y minutes)

I don't think this is precisely correct: hashes are remembered for `0 <= t <= X + Y` minutes. Or, put another way, until `t + Y < now`

Reply:
How about this?
```
// - the request's hash must not have been previously observed (hashes are remembered until they are Y in the past)
```